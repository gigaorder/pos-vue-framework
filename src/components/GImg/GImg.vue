<template>
    <div v-intersect:[intersectOptions].once="init" class="g-image" :style="containerStyles">
        <div v-if="loaded" class="g-image--content" :style="imageStyles">
            <slot></slot>
        </div>
        <template v-else>
            <slot name="placeholder">
                <div class="g-image--placeholder" :style="placeholderStyles">
                    <h3>Loading...</h3>
                </div>
            </slot>
        </template>
    </div>
</template>

<script>
  import {convertToUnit} from '../../utils/helpers';
  import {computed, onMounted, reactive, ref, watch} from '@vue/composition-api';
  import Intersect from "../../directives/intersect/intersect";

  export default {
    name: "GImg",

    directives: {
      Intersect
    },

    props: {
      src: [String, Object],
      lazySrc: [String],
      width: [Number, String],
      maxWidth: [Number, String],
      minWidth: [Number, String],
      height: [Number, String],
      maxHeight: [Number, String],
      minHeight: [Number, String],
      contain: Boolean,
      aspectRatio: [Number, String],
      gradient: String,
      intersectOptions: {
        type: Object,
        default: () => ({
          root: undefined,
          rootMargin: undefined,
          threshold: 0,
        }),
      },
    },

    setup(props, context) {
      let loaded = ref(false)

      const state = reactive({
        currentSrc: null,
        image: null,
        naturalWidth: 0,
        calculatedAspectRatio: 0
      })

      const backgroundImage = computed(() => {
        if (!(normalizedSrc.value.src || normalizedSrc.value.lazySrc)) return []

        const backgroundImages = []
        const src = !loaded.value ? normalizedSrc.lazySrc : state.currentSrc

        if (props.gradient) backgroundImages.push(`linear-gradient(${props.gradient})`)
        if (src) backgroundImages.push(`url("${src}")`)

        return backgroundImages.join(', ')
      })

      const normalizedSrc = computed(() => {
        return typeof props.src === 'string'
            ? {
              src: props.src,
              srcset: props.srcset,
              lazySrc: props.lazySrc,
              aspect: Number(props.aspectRatio),
            } : {
              src: props.src.src,
              srcset: props.srcset || props.src.srcset,
              lazySrc: props.lazySrc || props.src.lazySrc,
              aspect: Number(props.aspectRatio || props.src.aspect),
            }
      })

      watch(() => props.src, (newVal) => {
        // init again if src changes
        if (loaded.value) init(undefined, undefined, true)
        else loadImage()
      }, { lazy: true })

      function init(entries, observer, isIntersecting) {
        if (!isIntersecting) return

        if (normalizedSrc.lazySrc) {
          const lazyImg = new Image()
          lazyImg.src = normalizedSrc.lazySrc
          pollForSize(lazyImg, null)
        }

        if (normalizedSrc.value.src) loadImage()
      }


      function loadImage() {
        //load image here
        const _image = new Image()
        state.image = _image

        _image.onload = () => {
          if (_image.decode) {
            _image.decode().catch((err) => {
              console.warn(
                  `Failed to decode image, trying to render anyway\n\n` +
                  `src: ${normalizedSrc.value.src}`
              )
            }).then(onLoad)
          } else {
            onLoad()
          }
        }
        // _image.onerror = onError()

        // bind src to image
        normalizedSrc.value.src && (_image.src = normalizedSrc.value.src)
        props.sizes && (_image.sizes = props.sizes)
        normalizedSrc.value.srcset && (_image.srcset = normalizedSrc.value.srcset)
        props.aspectRatio || pollForSize(_image)
      }

      function onLoad() {
        getSrc()
        loaded.value = true
        context.emit('load', props.src)
      }

      function onError() {
        //fires error on Image() constructor
        console.error(`Image load failed\n\n` +
            `src: ${normalizedSrc.src}`)
        context.emit('error', props.src)
      }

      function getSrc() {
        if (state.image) state.currentSrc = state.image.currentSrc || state.image.src
      }

      // poll to calculate aspect ratio based on image sizes
      function pollForSize(image, timeout = 100) {
        const poll = () => {
          const {naturalHeight, naturalWidth} = image

          if (naturalHeight || naturalWidth) {
            state.naturalWidth = naturalWidth
            state.calculatedAspectRatio = naturalWidth / naturalHeight
          } else {
            timeout != null && setTimeout(poll, timeout)
          }
        }
        poll()
      }

      const containerStyles = computed(() => ({
        width: convertToUnit(props.width || (props.height * props.aspectRatio) || state.image.width),
        maxWidth: convertToUnit(props.maxWidth),
        minWidth: convertToUnit(props.minWidth),
        height: convertToUnit(props.height || (props.width / props.aspectRatio) || ((state.image.width / props.aspectRatio)) || state.image.height),
        maxHeight: convertToUnit(props.maxHeight),
        minHeight: convertToUnit(props.minHeight),
        aspectRatio: props.aspectRatio,
      }))

      const imageStyles = computed(() => {
        let contain = props.contain ? 'contain' : 'cover'
        return {
          backgroundImage: backgroundImage.value,
          backgroundSize: contain,
        }
      })

      const placeholderStyles = computed(() => {
        return {
          width: '100%',
          height: '100%',
          backgroundImage: 'url(https://picsum.photos/id/11/100/60)',
          backgroundRepeat: 'no-repeat',
          backgroundSize: 'contain',
          backgroundPosition: 'center center'
        }
      })

      return {
        containerStyles,
        imageStyles,
        placeholderStyles,
        loaded,
        normalizedSrc,
        backgroundImage,
        init,
      }
    }
  }
</script>
