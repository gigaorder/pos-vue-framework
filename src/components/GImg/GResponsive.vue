<template>
    <div></div>
</template>

<script>
    export default {
        name: "GResponsive",
        props: {
            height: [String, Number],
            maxHeight: [String, Number],
            minHeight: [String, Number],
            width: [String, Number],
            maxWidth: [String, Number],
            minWidth: [String, Number]
        }
    }
</script>

<style scoped>

</style>
