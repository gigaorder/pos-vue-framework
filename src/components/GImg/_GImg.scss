.g-image {
  position: relative;
  z-index: 0;

  &-content {
    position: absolute;
    z-index: -1;
    width: 100%;
    height: 100%;
    background-repeat: no-repeat;
    background-position: center center;
    background-color: #cccccc;
  }

//@media only screen and (max-width: 800px) {
//  &-content {
//    position: absolute;
//    z-index: -1;
//    width: 50%;
//    height: 50%;
//    background-repeat: no-repeat;
//    background-position: center center;
//    background-color: #cccccc;
//  }
//}

  &-placeholder {
    position: absolute;
    z-index: -1;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
  }
}

//.g-image {
//  z-index: 0;
//}
//
//.g-image__image,
//.g-image__placeholder {
//  z-index: -1;
//  position: absolute;
//  top: 0;
//  left: 0;
//  width: 100%;
//  height: 100%;
//
//  .g-image__image {
//    background-repeat: no-repeat;
//  }
//
//  &--preload {
//    filter: blur(2px)
//  }
//
//  &--contain {
//    background-size: contain
//  }
//
//  &--cover {
//    background-size: cover
//  }
//}
