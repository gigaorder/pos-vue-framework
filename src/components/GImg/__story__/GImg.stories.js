import {boolean, number, text, withKnobs} from '@storybook/addon-knobs';
import {action} from '@storybook/addon-actions';
import GImg from '../GImg';

export default {
  title: 'GImg',
  decorators: [withKnobs],
};

export const imgBasic = () => ({
  components: {GImg},
  prop:{
    contain:{default: boolean('contain',false)}
  },
  template: `<g-img src="https://i.ytimg.com/vi/B5Qb12DqZTA/maxresdefault.jpg"
               gradient="to top right, rgba(100,115,201,.33), rgba(255,32,72,.7)"
               height="400" aspect-ratio="1" :contain="contain"></g-img>`,
})

export const withHeightContain = () => ({
  components: {GImg},
  props: {
    contain: {default: boolean('contain', false)},
    height: {default: number('height', 500)},
    width: {default: number('width', 500)},
    aspectRatio: {default: number('ratio', null)},
  },
  template: `<g-img src="https://i.ytimg.com/vi/B5Qb12DqZTA/maxresdefault.jpg" :width="width" :height="height" :aspectRatio="aspectRatio" :contain="contain"
              :gradient="gradient"></g-img>`,
})

export const gradient = () => ({
  components: {GImg},
  props: {
    gradient: {default: text('gradient', 'to top right, rgba(100,115,201,.33), rgba(25,32,72,.7)')}
  },
  template: `<g-img src="https://i.ytimg.com/vi/B5Qb12DqZTA/maxresdefault.jpg" width="500" :height="500" contain :gradient="gradient"></g-img>`,
})

export const placeHolder = () => ({
  components: {GImg},
  props: {
    contain: {default: boolean('contain', false)},
    height: {default: number('height', 500)},
    width: {default: number('width', 500)},
  },
  template: `<div>
             <g-img src="http://www.effigis.com/wp-content/uploads/2015/02/Airbus_Pleiades_50cm_8bit_RGB_Yogyakarta.jpg" 
              lazy-src="https://picsum.photos/id/11/10/6"
              :width="width" :height="height" :aspectRatio="aspectRatio" :contain="contain">
               <template v-slot:placeholder><p>This is showed in loading time </p></template>
               <p style="background-color: #cccccc"> This is default placeholder </p>
             </g-img>
             <br>
             <g-img src="http://www.effigis.com/wp-_50cm_8bit_RGB_Yogyakarta.jpg" 
              lazy-src="https://picsum.photos/id/11/10/6"
              :width="width" :height="height" :aspectRatio="aspectRatio" :contain="contain">
               <template v-slot:placeholder><p>This is showed in loading time </p></template>
               <p style="background-color: #cccccc"> This is default placeholder </p>
             </g-img>
             </div>`,
})

export const lazyLoading = () => ({
  components: {GImg},
  props: {
    contain: {default: boolean('contain', false)},
    height: {default: number('height', 500)},
    width: {default: number('width', 500)},
  },
  template: `<div>
             <g-img src="http://www.effigis.com/wp-content/uploads/2015/02/Airbus_Pleiades_50cm_8bit_RGB_Yogyakarta.jpg" 
              lazy-src="https://picsum.photos/id/11/10/6"
              :width="width" :height="height" :aspectRatio="aspectRatio" :contain="contain">
               <template v-slot:placeholder><p>This is showed in loading time </p></template>
               <p style="background-color: #cccccc"> This is default placeholder </p>
             </g-img>
             <div style="height: 2000px; width: 500px"></div>
             <g-img src="http://www.effigis.com/wp-content/uploads/2015/02/Airbus_Pleiades_50cm_8bit_RGB_Yogyakarta.jpg" 
              lazy-src="https://picsum.photos/id/11/10/6"
              :width="width" :height="height" :aspectRatio="aspectRatio" :contain="contain">
               <template v-slot:placeholder><p>This is showed in loading time </p></template>
               <p style="background-color: #cccccc"> This is default placeholder </p>
             </g-img>
             </div>`,
})

export const srcset = () => ({
  components: {GImg},
  template: `<g-img src="https://i.ytimg.com/vi/B5Qb12DqZTA/maxresdefault.jpg"
                    srtset="https://i.ytimg.com/vi/B5Qb12DqZTA/maxresdefault.jpg 1000w,
                            https://picsum.photos/id/11/10/6 600w"
                    height="500px" width="500px"></g-img>`
})

// import Vue from 'vue/dist/vue.common.js'
// import {value} from "@vue/composition-api/dist/functions/state";
//
// describe('GImg', function () {
//   it('with width height', function () {
//     const vm = new Vue(imgWithWidthHeight()).$mount();
//     expect(vm.$el.outerHTML).toMatchSnapshot()
//   });
// })


