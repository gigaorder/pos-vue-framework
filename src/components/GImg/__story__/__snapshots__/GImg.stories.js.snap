// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GImg with width height 1`] = `
<div>
  <div class="g-image"
       style="background-size: contain; width: 500px; height: 500px; background-image: url(https://i.ytimg.com/vi/B5Qb12DqZTA/maxresdefault.jpg);"
  >
  </div>
  <div class="g-placeholder"
       style="background-size: contain; width: 500px; height: 500px;"
  >
  </div>
</div>
`;
